### ISSUE TITLE
neovim: plugin "vim-addon-nix" doesn't work properly

### ISSUE THREAD

**Issue by steveeJ** [nix-community/home-manager#283]
From *home.nix*:

```nix
  programs.neovim = {
    enable = true;
    configure = {
      packages.myVimPackage = with pkgs.vimPlugins; {
        start = [ vim-addon-nix ];
        opt = [ ];
      };
    };
  };
```

Results in the following error when trying to edit a nix file:

```
Error detected while processing /nix/store/cmxnqwx1wkhyknpj330qvhczyvpr7ax4-vimplugin-vim-addon-nix-2017-09-11/share/vim-plugins/vim-addon-nix/plugin/vim-addon-nix.vim:
line    8:
E117: Unknown function: funcref#Function
E15: Invalid expression: funcref#Function('vim_addon_nix#TagBasedCompletion')
line    9:
E117: Unknown function: funcref#Function
E15: Invalid expression: funcref#Function('vim_addon_nix#BuiltinsCompletion')
line   15:
E117: Unknown function: funcref#Function
E116: Invalid arguments for function actions#AddAction
Press ENTER or type command to continue

"~/.config/nixpkgs/home.nix" 605L, 12853C
Error detected while processing /nix/store/cmxnqwx1wkhyknpj330qvhczyvpr7ax4-vimplugin-vim-addon-nix-2017-09-11/share/vim-plugins/vim-addon-ni
x/ftplugin/nix.vim:
line   14:
E117: Unknown function: funcref#Function
E116: Invalid arguments for function on_thing_handler#AddOnThingHandler
```

**Comment by rycee** [comment]
Sorry, I don't have the faintest idea what might cause this 😕

@teto Maybe you have some idea?

**Comment by steveeJ** [comment]
I was pointed towards this on IRC so I don't fully understand what is wrong, but I have found a solution.
The plugin installation as described above doesn't work with all plugins, using *vam* instead does. 
The above example looks like this when using the *vam* method:

```nix
  programs.neovim = {
    enable = true;
    configure = {
      vam.knownPlugins = pkgs.vimPlugins; # optional
      vam.pluginDictionaries = [
        { names = [ "vim-addon-nix" ]; }
      ];
    };
  };
```

As this is not specific to home-manager, more information is already documented here https://nixos.wiki/wiki/Vim#Using_VAM_as_manager

Thanks for taking a look @rycee.

I'll leave the issue open until we've figured out that the default/recommended way of installing (neo)vim plugins with home-manager uses vam or finds a different solution that supposedly works with all plugins.

**Comment by teto** [comment]
The configure system of (neo)vim should be changed for something easier to read (the nix code is really hard to follow), run some check phase to test that the init.vim doesn't have error, to account for the different paths between vim and neovim etc.
Just need volunteers :D

**Comment by BadDecisionsAlex** [comment]
I honestly had to completely scrap using Nix/NixOS/Home-Manager for my Neovim plugins. I do them the ol' fashioned way and occasionally drop store path expansions as variables in a tiny .vim file that my init.vim imports.
:(

This is still not without issues. I don't have a clue how to update the instance of node that NeoVim is using.

**Comment by kalbasit** [comment]
> I honestly had to completely scrap using Nix/NixOS/Home-Manager for my Neovim plugins. I do them the ol' fashioned way and occasionally drop store path expansions as variables in a tiny .vim file that my init.vim imports.

I've been using Nix to manage Vim since day 1 and I never really had any issues. Although, Nix does comes with 3 different managers for plugins and it's quite confusing. But I did find VAM capable.

I have [a separate NeoVim module](https://github.com/kalbasit/shabka/tree/master/modules/neovim) that both home-manager and nixos are using to install/configure it system-wide (on nixos) and for the user (on Darwin).

**Comment by teto** [comment]
I have a pending PR in nixpkgs to improve its configuration but overall it works well.
@BadDecisionsAlex nodejs is passed via the wrapper. You could use `wrapNeovim.override({nodejs= ....})` to pass your own version of node

**Comment by teto** [comment]
Old, feel free to reopen
